### PR TITLE
feat: add tari_crypto hashing api support

### PR DESCRIPTION
**Pull request by hansieodendaal**
Description
---
Added a hashing domain API for general use based on the new domain separated hashing API from `tari_crypto`; this is a proposal to implement the hashing domain API system wide. A new struct `HashingDomain` was added to provide a standardized way of implementing the new tari_crypto hashing api, which provides standardized domain separated hashing, with two tari project specific hash domain structs, `DefaultHashDomain` and `MacHashDomain`.

USe of `HashingDomain` allows differentiated domain declarations per "hash originating generation" crate level (_see **Suggested hashing domains** below_). The domain labels were chosen to complement the tari project specific hash domain label prefixes provided by `DefaultHashDomain` and `MacHashDomain` as `"com.tari.tari_project.hash_domain.v1"` and  `"com.tari.tari_project.mac_domain.v1"`. As an example, resulting domain labels for `COMMON_HASH_DOMAIN` that specifies its own label `"common"` will then be `"com.tari.tari_project.hash_domain.v1.common"` and `"com.tari.tari_project.mac_domain.v1.common"`.

An example use case is:
``` rust
        let mut hasher = common_hash_domain().hasher::<Blake256>();
        hasher.update(b"my 1st secret");
        hasher.update(b"my 2nd secret");
        let hash = hasher.finalize();
```

The general idea is to use the lowest level hashing domain for the crate where an original hash needs to be generated, and where no higher level hash domain is defined for a specific crate,  `common_hash_domain()` needs to be used. Code that replays a specific hash needs to use the same hash domain where the hash originated, for example the mempool. 

This standardized domain declarations should make code maintenance where domain hashing is used simpler and more efficient. The next PR will demonstrate use of `tari_script_hash_domain()` and `mmr_hash_domain()`.

**Suggested hashing domains** (_This can easily be modified as needed._)
``` rust
pub fn common_hash_domain() -> HashingDomain {HashingDomain::new("common")}
pub fn comms_core_hash_domain() -> HashingDomain {HashingDomain::new("comms.core")}
pub fn comms_dht_hash_domain() -> HashingDomain {HashingDomain::new("comms.dht")}
pub fn core_hash_domain() -> HashingDomain {HashingDomain::new("base_layer.core")}
pub fn core_blocks_hash_domain() -> HashingDomain {HashingDomain::new("base_layer.core.blocks")}
pub fn core_consensus_hash_domain() -> HashingDomain {HashingDomain::new("base_layer.core.consensus")}
pub fn core_covenants_hash_domain() -> HashingDomain {HashingDomain::new("base_layer.core.covenants")}
pub fn core_proof_of_work_hash_domain() -> HashingDomain {HashingDomain::new("base_layer.core.proof_of_work")}
pub fn core_transactions_hash_domain() -> HashingDomain {HashingDomain::new("base_layer.core.transactions")}
pub fn dan_layer_core_hash_domain() -> HashingDomain {HashingDomain::new("dan_layer.core")}
pub fn dan_layer_engine_hash_domain() -> HashingDomain {HashingDomain::new("dan_layer.engine")}
pub fn dan_layer_hash_domain() -> HashingDomain {HashingDomain::new("dan_layer")}
pub fn key_manager_hash_domain() -> HashingDomain {HashingDomain::new("base_layer.key_manager")}
pub fn mmr_hash_domain() -> HashingDomain {HashingDomain::new("base_layer.mmr")}
pub fn p2p_hash_domain() -> HashingDomain {HashingDomain::new("base_layer.p2p")}
pub fn tari_script_hash_domain() -> HashingDomain {HashingDomain::new("infrastructure.tari_script")}
```

Motivation and Context
---
Needed to implement the `tari_crypto` standardized domain separated hashing.

How Has This Been Tested?
---
Unit tests (_no system level impact with this PR yet_).

### CODE REVIEW

File: common/src/hashing_domain.rs [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+        GenericDomainHasher::new(self.domain_label)
+    }
+
+    /// Convenience function to compute hash of the data. It will handle hasher creation, data feeding and finalization.
+    pub fn digest<D: Digest>(&self, data: &[u8]) -> DomainSeparatedHash {
+        self.hasher::<D>().chain(data).finalize()
+    }
+
+    /// Helper function to return a fixed 32 size hashed bytes array
+    pub fn hash_to_bytes_32(hash: &DomainSeparatedHash) -> Result<[u8; 32], HashingDomainError> {
```
Comment (therustmonk) [inline common/src/hashing_domain.rs]: How about to add a trait for that and implement it for the `DomainSeparatedHash`? [[demo](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=a82d0ad66d95d484a7d809363912a9aa)]

```rust
pub trait HashToBytes<const I: usize>: AsRef<[u8]> {
    fn hash_to_bytes(&self) -> Result<[u8; I], HashingDomainError> {
        let hash_vec = self.as_ref();
        if hash_vec.is_empty() || hash_vec.len() < I {
            return Err(HashingDomainError::CopyFromSlice);
        }
        let mut buffer: [u8; I] = [0; I];
        buffer.copy_from_slice(&hash_vec[..I]);
        Ok(buffer)
    }
}

// It has to be implemented for the `DomainSeparatedHash` instead
impl<const I: usize> HashToBytes<I> for Vec<u8> {}

pub fn somewhere() -> Result<(), Error> {
    let data = vec![0u8,1,2,3,4,5,6,7,8];
    let bytes: [u8; 8] = data.hash_to_bytes()?;
    let bytes: [u8; 4] = data.hash_to_bytes()?;

    // Error is thrown here (not enough bytes the slice has)
    let bytes: [u8; 16] = data.hash_to_bytes()?;
    Ok(())
}
```

File: common/src/hashing_domain.rs [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+
+pub struct HashingDomain {
+    domain_label: &'static str,
+}
+
+/// Error type for the pipeline.
+#[derive(Debug, Error)]
+pub enum HashingDomainError {
+    #[error("Two slices have different lengths")]
+    CopyFromSlice,
```
Comment (stringhandler) [inline common/src/hashing_domain.rs:36]:
```suggestion
    SlicesHaveDifferentLengths
```

File: common/src/lib.rs [hunk 3]
```diff
@@ -76,3 +80,11 @@ pub const DEFAULT_MINER_LOG_CONFIG: &str = "config/log4rs_miner.yml";
 pub const DEFAULT_COLLECTIBLES_LOG_CONFIG: &str = "config/log4rs_collectibles.yml";
 
 pub(crate) const LOG_TARGET: &str = "common::config";
+
+/// The MMR domain separated hashing domain
```
Comment (stringhandler) [inline common/src/lib.rs:84]: This common is wrong

**Review by stringhandler — APPROVED** [review]
I'm not sure this is exactly what @CjS77 intended, but let's see what it looks like when you have replaced more functionality with the hashers

**Pull request merged**